File: src/RayTracing/Solver.ts
const SolveQuadratic = (a: number, b: number, c: number): null | number | [number, number] => {
  const determinant = (b ** 2) - (4 * a * c);

  if (determinant < 0) return null;

  if (determinant === 0) {
    return (-b) / (2 * a);
  }

  const plus = (-b + Math.sqrt(determinant)) / (2 * a);
  const minus = (-b - Math.sqrt(determinant)) / (2 * a);

  return [plus, minus];
};

// there might be more here in the future
// eslint-disable-next-line import/prefer-default-export
export { SolveQuadratic };

File: src/RayTracing/rs-ray-tracing/src/solver.rs
pub fn solve_quadratic (a: f64, b: f64, c: f64) -> Option<(f64, f64)> {
  let determinant = b.powi(2) - (4. * a * c);

  if determinant < 0. {
    return None;
  }

  if determinant == 0. {
    let solution = (-b) / (2. * a);
    return Some((solution, solution));
  }

  let plus = (-b + determinant.sqrt()) / (2. * a);
  let minus = (-b - determinant.sqrt()) / (2. * a);

  Some((plus, minus))
}